set extension browser timeout to minimum of 10 seconds.

MegaphoneJon opened the pull request:
Overview
----------------------------------------
The default timeout waiting for a Guzzle request to respond is 5 seconds.  The "extdir" app at https://civicrm.org/extdir sometimes takes more than 5 seconds to respond if the response isn't cached.  This leads to false positives in the CiviCRM extension check.

For more discussion, see https://lab.civicrm.org/infra/ops/-/issues/1009.

Before
----------------------------------------
Guzzle timeout comes from `http_timeout` setting (defaults to 5 seconds).

After
----------------------------------------
Guzzle timeout has a minimum of 10 seconds in `CRM_Extension_Browser`.

MegaphoneJon committed:
set extension browser timeout to minimum of 10 seconds.

colemanw commented:
Thanks @MegaphoneJon. Perhaps we should just bump the default value here?
https://github.com/civicrm/civicrm-core/blob/e9ccbb7169fabc0fe91e8187dfddcc3b4f8ec959/settings/Core.setting.php#L1196

MegaphoneJon commented:
I considered that - but this value is used elsewhere.  And while I've occasionally been bitten by some cURL request taking too long, I wasn't sure what the ramifications would be elsewhere in Civi.  For instance, I believe the "CiviCRM News" dashlet relies on this - I'd rather have that time out sooner so it doesn't look so obviously broken to the casual end user.

agileware-justin commented:
Thanks @eileenmcnaughton 

I have many questions!

Why 10 seconds as a forced minimum? When you can simply increase the http_timeout.

Why is this check so important that the system should wait for it to complete? When another system check may complete within the time limit.

Why place an external dependency in a system check at all? What about sites which have no connection to the Internet or are otherwise limited.

In my experience system checks are performed far too frequently as it is and at the most inopportune times, so potentially slowing this down further will only increase user frustration.

eileenmcnaughton commented:
My understanding - @MegaphoneJon can confirm - is this is the guzzle usage from the extensions page rather than a system check

agileware-justin commented:
Can we also remove this line? Appears redundant - can't see it being used.

` const CHECK_TIMEOUT = 5;`

https://github.com/civicrm/civicrm-core/pull/27934/files#diff-40007add95a7a02a22b0db45518b6541d4105a619f64ce810e1eafc74c68fb58R40

MegaphoneJon commented:
@agileware-justin I removed that CONST, thanks for pointing it out.

I checked, and this function is called by the extension check.  So it's worth addressing your points.
>  Why 10 seconds as a forced minimum? When you can simply increase the http_timeout.

It's covered in more depth on the [lab.c.o ticket](https://lab.civicrm.org/infra/ops/-/issues/1009.) but: The extension JSON takes > 5 seconds to generate. So users see an intermittent error.

I could increase *my* HTTP timeout, but I want to avoid Civi users at large getting an error here. 
  
>  Why is this check so important that the system should wait for it to complete? When another system check may complete within the time limit.

It's not more important than others - but it's the only one that generates an error under normal circumstances.
  
>  Why place an external dependency in a system check at all? What about sites which have no connection to the Internet or are otherwise limited.

This would eliminate some of the most important checks, like "security update available".  This particular check can be disabled by specifying no external repo, which is the correct configuration for sites without Internet.  Otherwise the Extensions page will be broken.
  
>  In my experience system checks are performed far too frequently as it is and at the most inopportune times, so potentially slowing this down further will only increase user frustration.

This is a wider discussion - I don't believe they're performed synchronously on page load but I could be wrong.  But I agree that not every check should run on `System.check` unless forced.  It should work more like Scheduled Jobs, where an admin can determine how frequently checks run.

Proposal:
* 2 new columns in `civicrm_status_pref`: `run_frequency` and `last_run` which performs the same task as the same field in Scheduled Jobs.  We should probably learn from the Scheduled Job extensions that give finer control. No specified `run_frequency` means "keep current behavior".
* A `force` argument to `System.check` to run all checks regardless of last run time.

agileware-justin commented:
Thanks for your considered comments @MegaphoneJon - thumbs up 👍🏻

eileenmcnaughton commented:
ok - thanks @agileware-justin I'll merge
